lib/Serialization/Deserialization.cpp

```diff
@@ -4552,24 +4552,34 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
         break;
       }
 
-      case decls_block::HasAsyncAlternative_DECL_ATTR: {
+      case decls_block::CompletionHandlerAsync_DECL_ATTR: {
         bool isCompound;
         ArrayRef<uint64_t> rawPieces;
-        serialization::decls_block::HasAsyncAlternativeDeclAttrLayout::readRecord(
-            scratch, isCompound, rawPieces);
-
-        DeclNameRef name;
+        uint64_t handlerIndex;
+        bool explicitHandlerIndex;
+        uint64_t asyncFunctionDeclID;
+        serialization::decls_block::CompletionHandlerAsyncDeclAttrLayout::
+            readRecord(scratch, explicitHandlerIndex, handlerIndex,
+                       asyncFunctionDeclID, isCompound, rawPieces);
+
+        DeclNameRef asyncFunctionName;
         if (!rawPieces.empty()) {
           auto baseName = MF.getDeclBaseName(rawPieces[0]);
           SmallVector<Identifier, 4> pieces;
           for (auto rawPiece : rawPieces.drop_front())
             pieces.push_back(MF.getIdentifier(rawPiece));
-          name = !isCompound ? DeclNameRef({baseName})
-                             : DeclNameRef({ctx, baseName, pieces});
+          asyncFunctionName = !isCompound
+                                  ? DeclNameRef({baseName})
+                                  : DeclNameRef({ctx, baseName, pieces});
         }
 
-        Attr = new (ctx) HasAsyncAlternativeAttr(name, SourceLoc(),
-                                                 SourceRange());
+        auto adf = cast<AbstractFunctionDecl>(MF.getDecl(asyncFunctionDeclID));
+        Attr = new (ctx) CompletionHandlerAsyncAttr(
+            asyncFunctionName, /*functionNameLoc*/ SourceLoc(),
+            explicitHandlerIndex, handlerIndex, /*handlerIndexLoc*/ SourceLoc(),
+            /*atLoc*/ SourceLoc(), /*range*/ SourceRange());
+        reinterpret_cast<CompletionHandlerAsyncAttr *>(Attr)
```

Comment:
Yeah, since I'm creating the thing on the statement above, we _know_ that the `Attr` variable contains that type.
The `reinterpret_cast` says that I know what I'm doing and want to cast the type without any conversions and is completely free. The `cast` on the other hand actually does runtime checking and isn't free.

That said, I think a `static_cast` will do the trick and would definitely be better than a yolo-cast.